### ISSUE TITLE
flowable segfaults after clearing and pushing less elements than on previous render

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#210]
Apply this patch and enjoy the crashes!

```diff
diff --git a/examples/layout/main.cpp b/examples/layout/main.cpp
index c97e5c3..0ed1c91 100644
--- a/examples/layout/main.cpp
+++ b/examples/layout/main.cpp
@@ -144,28 +144,37 @@ auto make_htile_mixed()
    );
 }

+auto num_elements = 400;
+
 auto make_flow()
 {
    constexpr auto line_height = 30;
    constexpr auto min_size = 10;
    constexpr auto max_width = 100;
    constexpr auto max_height = line_height;
-   constexpr auto num_elements = 40;

    static auto c = flow_composite{};
-   for (int i = 0; i < num_elements; ++i)
-   {
-      auto w = min_size + ((double(std::rand()) * (max_width - min_size)) / RAND_MAX);
-      auto h = min_size + ((double(std::rand()) * (max_height - min_size)) / RAND_MAX);
-      auto _box = vsize(line_height, align_bottom(margin(
-         { 5, 5, 5, 5 }, fixed_size({ float(w), float(h) }, rbox_)
-      )));
-      c.push_back(share(_box));
-   }
+
+   auto btn = button("new tiles");
+   btn.on_click = [&](bool) {
+      c.clear();
+      for (int i = 0; i < num_elements; ++i)
+      {
+        auto w = min_size + ((double(std::rand()) * (max_width - min_size)) / RAND_MAX);
+        auto h = min_size + ((double(std::rand()) * (max_height - min_size)) / RAND_MAX);
+        auto _box = vsize(line_height, align_bottom(margin(
+          { 5, 5, 5, 5 }, fixed_size({ float(w), float(h) }, rbox_)
+        )));
+        c.push_back(share(_box));
+      }
+      num_elements -= 150;
+   };
+
+   btn.on_click(false);

    auto flow_pane = margin(
       { 0, 50, 10, 10 },
-      align_top(flow(c))
+      vtile(btn, align_top(flow(c)))
    );

    return margin({ 10, 10, 10, 10 },
```

Now, to the actual issue, here are my observations:

The segfault occurs in `htile::limits` on line `auto el = at(i).limits(ctx);`. There is a broken lifetime:

```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x00000000004409b6 in cycfi::elements::htile_element::limits (this=0xd209d90,
    ctx=...)
    at [...]/lib/src/element/tile.cpp:182
182              auto el = at(i).limits(ctx);
(gdb) print this->size()
$1 = 2
(gdb) print i
$2 = 1
(gdb) print this->at(0)
$3 = (cycfi::elements::element &) @0xa81eeb0: {<std::enable_shared_from_this<cycfi::elements::element>> = {
    _M_weak_this = {<std::__weak_ptr<cycfi::elements::element, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0xa81eeb0, _M_refcount = {
          _M_pi = 0xa81eea0}}, <No data fields>}},
  _vptr.element = 0x4c7b10 <vtable for item_preview+16>}
(gdb) print this->at(1)
$4 = (cycfi::elements::element &) @0xa823150: {<std::enable_shared_from_this<cycfi::elements::element>> = {
    _M_weak_this = {<std::__weak_ptr<cycfi::elements::element, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0xfeeefeeefeeefeee, _M_refcount = {
          _M_pi = 0xfeeefeeefeeefeee}}, <No data fields>}},
  _vptr.element = 0xfeeefeeefeeefeee}
```
The element at index `1` is already dead at this point.

I strongly suspect this code:

https://github.com/cycfi/elements/blob/1f5ba55f6f5222eb1585f431499b29f145916daa/lib/src/element/flow.cpp#L64-L68

For each row, it creates an instance of range composite which only holds a reference to the container and copies of the indexes. I think this breaks when flowable container contents are changed, if there are less elements in flowable later then this causes access out of bounds because new flow has less rows.

**Comment by djowel** [comment]
OK... so... again, this is not a bug. The least I can say is that it is a documentation bug, for lack of info...

When you change a composite, you need to ask the view or a parent element, to layout its contents. This is true of all the containers, such as vtile, flow, htile, etc. It makes sense. With the modular nature of elements, a composite does not generally know if its children changed. If you steal or add one or more of its children, you have to inform it :-). You need to notify it of the change. It may be possible to do it automatically by maintaining a checksum and doing the check every time but that is another matter open for discussion.

In this case it is simplest to just add:

```c++
view_.layout();
```

in your callback that changes the number of items.

P.S. I know better if an "issue" is a bug or not. May I request that I do the 'bug' tagging, or maybe consult with me first? Also, may I request you to issue another issue on the documentation of re-layouting composites? Thanks!

**Comment by Xeverous** [comment]
> You need to notify it of the change. It may be possible to do it automatically by maintaining a checksum and doing the check every time but that is another matter open for discussion.

IMO this is another point to discuss for #209.

You said earlier that GUI should react nicely to user faults, which is the case for sliders which clamp the value. I find this to be a counter-intuitive example because a missing layout action causes a crash.

> In this case it is simplest to just add:
>
>     view_.layout();
>
> in your callback that changes the number of items.

No, it is not simple:

- I don't pass `view` along my code which creates UI.
- If I need to call view's layout, this means that I need to have view before creating/binding my UI.
- If I need to call anything on view, this means I need to construct a window first because window handle is required by view's constructor.

In short, there is a strong coupling for things which should not need each other.

IMO it would be far better if such things (layout, refresh) were done on every frame. This decouples element-creation code from OS-level wrappers and for any sufficiently complex application manually tracking UI changes is not worth it since it will almost something (or multiple elements) will be changed on every frame.

You haven't replied in #209, but - would you be interested in the possibility of a more explicit core-event-update-render-loop code? Soon I may have some time to experiment with different OS abstraction implementations and you probably know that I'm not very satisfied with current implementation which hides everything inside.

> I know better if an "issue" is a bug or not. May I request that I do the 'bug' tagging, or maybe consult with me first?

This time I was sure it was a bug. I did not expect the program to crash just by changing flowable contents. From user's perspective, it is a bug because nothing ever mentions the need for layout call. Documentation bugs are also bugs. Either way, something is incorrect.

But ok, I won't tag if you don't want to. I will just let you know that many repositories are using issue templates now so the tags are automatically added when the issue is created because a user choose "bug issue template" or so. Some of such repositories often have multiple bug-related tags: `bug unconfimed`, `bug confirmed`, `bug can't reproduce` etc.

> may I request you to add another issue on the documentation of re-layouting composites? Thanks!

If you think that missing documentation is the cause of this crash, then IMO *this* is the issue about missing information in the documentation. I would just recommend to change the title.

**Comment by djowel** [comment]
OK, I pushed the winning solution to develop. 

Rather tricky because there are two data structures where one (flow tiles) should sync with the other (children elements). 

The winning solution requires the user to notify the flow_composite of any change via `_flow_composite.reflow();` before `view_.layout();`.

- I tried other solutions that does not require notification, but this is the safest and most efficient. Only the user knows when the composite changed. 
- It is possible to detect changes programmatically, but I always think thousands or even millions of elements and that will be expensive. While OTOH the user changing the elements knows when such changes happen. 
- It is also possible to overload the mutating member functions such as push_back and track the change on such calls, but that is inflexible and not 100% fool proof (e.g. operator[] is not necessarily mutation, neither are calls to begin() end(), front(), beck(), etc.).
